loki/transformations/utilities.py

```diff
@@ -32,10 +35,167 @@
+        combine_map = {routine_args[0]: as_tuple(routine_args[1:]) for routine_args in combine}
+        arg_map = {arg.name: rename_common_map[common_arg.name]
+                if common_arg.name in rename_common_map and rename_common_map[common_arg.name] is not None
+                else common_arg.name
+                for common_arg, redundant_args in combine_map.items() for arg in redundant_args}
+        # remove duplicates from callee.arguments
+        new_routine_args = tuple(arg for arg in callee.arguments if arg not in redundant)
+        # rename if common name is possible
+        new_routine_args = as_tuple(arg.clone(name=rename_common_map[arg.name])
+                if arg.name in rename_common_map else arg for arg in new_routine_args)
+        callee.arguments = new_routine_args
+
+        # rename usage/occurences in callee.body
+        var_map = {}
+        variables = FindVariables(unique=False).visit(callee.body)
+        var_map = {var: var.clone(name=arg_map[var.name]) for var in variables if var.name in arg_map}
+        var_map.update({var: var.clone(name=rename_common_map[var.name]) for var in variables
+            if var.name in rename_common_map})
+        callee.body = SubstituteExpressions(var_map).visit(callee.body)
+        # modify the variable declarations, thus remove redundant variable declarations and possibly rename
+        modify_variable_declarations(callee, remove_symbols=redundant, rename_symbols=rename_common_map)
+        # store the information for possibly later renaming kwarguments on caller side
+        return rename_common_map
+
+    def rename_kwarguments(relevant_calls, rename_common_map_routine):
+        for call in relevant_calls:
+            kwarguments = call.kwarguments
+            if kwarguments:
+                call_name = str(call.routine.name).lower()
+                new_kwargs = as_tuple((rename_common_map_routine[call_name][kw[0]], kw[1])
+                        if kw[0] in rename_common_map_routine[call_name] else kw for kw in kwarguments)
+                call._update(kwarguments=new_kwargs)
+
+    calls = FindNodes(CallStatement).visit(routine.body)
+    call_arg_map = {}
+    relevant_calls = []
+    # adapt call statements (and remove duplicate args/kwargs)
+    for call in calls:
+        if call.routine is BasicType.DEFERRED:
+            continue
+        call_arg_map[call.routine] = remove_duplicate_args_call(call)
+        relevant_calls.append(call)
+    rename_common_map_routine = {}
+    # modify/adapt callees
+    for callee, callee_arg_map in call_arg_map.items():
+        rename_common_map_routine[str(callee.name).lower()] = modify_callee(callee, callee_arg_map)
+    # handle possibly renamed kwarguments on caller side
+    if rename_common:
+        rename_kwarguments(relevant_calls, rename_common_map_routine)
+
+
+def modify_variable_declarations(routine, remove_symbols=(), rename_symbols=None):
```

Comment:
This partially re-implements the setter for `ProgramUnit,.variables`: https://github.com/ecmwf-ifs/loki/blob/cc179b3b86613af86df6dabb0be250cde4414fc9/loki/program_unit.py#L505

Should we maybe make this functionality a method on `ProgramUnit` and use it in both cases? (Cc @mlange05 for input)